nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/toolkit/tls/service/TlsCertificateAuthorityClient.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    public void generateCertificateAndGetItSigned() throws Exception {
+        generateCertificateAndGetItSigned(null);
+    }
+
+    public void generateCertificateAndGetItSigned(String certificateFile) throws Exception {
+        KeyPair keyPair = tlsHelper.generateKeyPair();
+
+        String keyStoreType = tlsClientConfig.getKeyStoreType();
+        if (StringUtils.isEmpty(keyStoreType)) {
+            keyStoreType = TlsConfig.DEFAULT_KEY_STORE_TYPE;
+            tlsClientConfig.setKeyStoreType(keyStoreType);
+        }
+
+        KeyStore keyStore = tlsHelper.createKeyStore(keyStoreType);
+        String keyPassword = tlsClientConfig.getKeyPassword();
+        char[] passphrase;
+        if (TlsHelper.PKCS12.equals(keyStoreType)) {
+            passphrase = null;
+            tlsClientConfig.setKeyPassword(null);
+        } else {
+            if (StringUtils.isEmpty(keyPassword)) {
+                keyPassword = passwordUtil.generatePassword();
+                tlsClientConfig.setKeyPassword(keyPassword);
+            }
+            passphrase = keyPassword.toCharArray();
+        }
+        X509Certificate[] certificates = tlsCertificateSigningRequestPerformer.perform(objectMapper, keyPair);
+        tlsHelper.addToKeyStore(keyStore, keyPair, TlsToolkitMain.NIFI_KEY, passphrase, certificates);
+
+        String keyStorePassword = tlsClientConfig.getKeyStorePassword();
+        if (StringUtils.isEmpty(keyStorePassword)) {
+            keyStorePassword = passwordUtil.generatePassword();
+            tlsClientConfig.setKeyStorePassword(keyStorePassword);
+        }
+
+        try (OutputStream outputStream = outputStreamFactory.create(new File(tlsClientConfig.getKeyStore()))) {
+            keyStore.store(outputStream, keyStorePassword.toCharArray());
+        }
+
+        String trustStoreType = tlsClientConfig.getTrustStoreType();
+        if (StringUtils.isEmpty(trustStoreType)) {
+            trustStoreType = TlsConfig.DEFAULT_KEY_STORE_TYPE;
+            tlsClientConfig.setTrustStoreType(trustStoreType);
+        }
+
+        KeyStore trustStore = tlsHelper.createKeyStore(trustStoreType);
+        trustStore.setCertificateEntry(TlsToolkitMain.NIFI_CERT, certificates[certificates.length - 1]);
+
+        String trustStorePassword = tlsClientConfig.getTrustStorePassword();
+        if (StringUtils.isEmpty(trustStorePassword)) {
+            trustStorePassword = passwordUtil.generatePassword();
+            tlsClientConfig.setTrustStorePassword(trustStorePassword);
+        }
+
+        try (OutputStream outputStream = outputStreamFactory.create(new File(tlsClientConfig.getTrustStore()))) {
+            trustStore.store(outputStream, trustStorePassword.toCharArray());
+        }
+
+        if (!StringUtils.isEmpty(certificateFile)) {
+            try (Writer writer = new OutputStreamWriter(outputStreamFactory.create(new File(certificateFile)))) {
```

Comment:
We might need to restrict the POSIX permissions for the certificates, keystore/truststore, key, and properties/config files being written out (not specific to this method).